kubessh/sftp.py

```diff
@@ -0,0 +1,164 @@
+
+    def _init_delegate(self):
+        self._proc = subprocess.Popen(
+            ['./sftp-server'],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=sys.stderr
+        )
+        self._proc_stdin = self._proc.stdin
+        self._proc_stdout = self._proc.stdout
+
+    async def _cleanup(self, exc):
+        self._proc.terminate()
+        await super()._cleanup(exc)
+
+    async def _forward_incoming_packet_to_delegate(self, data):
+        self.logger.debug1(f'write {len(data)} byte packet to delegate process')
+        self.logger.debug2(f'{data}')
+        self._proc_stdin.write(
+            len(data).to_bytes(length=4, byteorder='big', signed=False))
+        self._proc_stdin.write(data)
+        self._proc_stdin.flush()
+
+    async def _read_packet_from_delegate(self):
+        self.logger.debug1('read packet from delegate process')
+        ch = self._proc_stdout
+        self.logger.debug2('read packet length field')
+        pkt_len = int.from_bytes(ch.peek()[:4], byteorder='big', signed=False)
+        self.logger.debug2(f'attempt to read {pkt_len} byte packet')
+        packet = ch.read(pkt_len + 4)
+        self.logger.debug2(f'delegate packet is {packet}')
+        return packet
+
+    async def _forward_outgoing_packet_from_delegate(self, data):
+        self._writer.write(data)
+
+    async def _process_packet(self, pkttype, pktid, packet):
+        self.log_received_packet(pkttype, pktid, packet, note='forwarded to subprocess')
+        await self._forward_incoming_packet_to_delegate(packet.get_full_payload())
+        response = await self._read_packet_from_delegate()
+        # TODO: (would have to redundantly decode packet to log it, omitted for now)
+        #self.log_sent_packet(pkttype, pktid, packet, note='forwarded from subprocess')
+        await self._forward_outgoing_packet_from_delegate(response)
+
+    async def run(self):
+        # skip super's SFTP init handling, the subprocess does this here
+        await self.recv_packets()
+
+
+class KubeSFTPServerHandler(DelegateSFTPServerHandler, LoggingConfigurable):
+
+    # where the SFTP binary lives in the KubeSSH server image
+    sftp_binary_source_path = Unicode(
+        '/usr/local/share/kubessh/sftp-server',
+        allow_none=False,
+        help="""
+        Path in the KubeSSH server container to the static sftp-server binary.
+        """,
+        config=True
+    )
+
+    # where the SFTP binary goes in the user pod
+    sftp_binary_upload_path = Unicode(
+        '.local/sbin/kubessh-sftp-server',
+        allow_none=False,
+        help="""
+        Path inside the user pod where the sftp-server binary will be uploaded.
+
+        Must be full path including the executable name, and can be absolute or
+        relative to the working dir. Must be writable by the user the pod runs
+        as.
+        """,
+        config=True
+    )
+
+    def __init__(self, server, reader, writer):
```

Comment:
Do you think you can write docstrings for each of these methods? Would be very helpful in deocoding what exactly is happening here.

Reply:
Sure, will do!